[WFLY-14365] Programmatic web authentication (HttpServletRequest.login()) does not trigger sso

darranl opened the pull request:
As with the PRs to WildFly Elytron and Elytron Web I am submitting this as a draft PR as there is some follow up work required which I will likely complete just after feature freeze but as clustered SSO is a sensitive area of code I would like to get reviews started in case a different strategy is needed.

This PR depends on the following PRs so will not compile without component upgrades:
  https://github.com/wildfly-security/wildfly-elytron/pull/1483
  https://github.com/wildfly-security/elytron-web/pull/184

The WildFly changes really cover two areas:
1. Enabling the use of the ProgrammaticSingleSignOnCache to activate SSO for programmatic authentication.
2. Adding an additional field to the replicated CachedIdentity to track if the identity was created as a result of programmatic authentication.

https://issues.redhat.com/browse/WFLY-14365

darranl committed:
[WFLY-14365] Provide a ProgrammaticSingleSignOnCache to enable SSO for programmatic authentication.

darranl committed:
[WFLY-14365] Switch to a BiFunction so the ProgrammaticSingleSignOnCache can be created with the mechanism name as well.

darranl committed:
[WFLY-14365] Update clustering SSO integration to match the Elytron API / SPI updates.

darranl commented:
I have submitted the following snapshot-ci runs:

The first run attempts to use the topic branches of WildFly and WildFly Elytron against the master branches of WildFly Core and WildFly - this is to check if the component upgrades can be applied cleanly before applying the WildFly changes.

https://github.com/wildfly/snapshot-ci/issues/12

This second run is the same as the first but uses a topic branch for WildFly with the final changes applied.

https://github.com/wildfly/snapshot-ci/issues/16

pferraro commented:
@darranl If you rebase this with master, I can submit a PR to your branch that adjusts the marshalling code accordingly.

darranl commented:
Thanks @pferraro I will do that first today.

darranl commented:
@pferraro I have cleaned up this PR.

I have fixed up the compilation failures / test failures I was seeing myself in a normal WildFly run, not sure what else is needed.

In the past I think you said there was an SSO test in the clustering testsuite - next week when I get some time I will try and adapt that to add a programmatic scenario.

darranl committed:
[WFLY-14365] Add "programmatic" to the protobuf schema.

darranl commented:
@pferraro I have added that field, and learnt the hard way all my editors are nicely configured to refuse to use tabs!!
